common/utils/include/claragenomics/utils/graph.hpp

```diff
@@ -0,0 +1,300 @@
+/*
+* Copyright (c) 2019, NVIDIA CORPORATION.  All rights reserved.
+*
+* NVIDIA CORPORATION and its licensors retain all intellectual property
+* and proprietary rights in and to this software, related documentation
+* and any modifications thereto.  Any use, reproduction, disclosure or
+* distribution of this software and related documentation without an express
+* license agreement from NVIDIA CORPORATION is strictly prohibited.
+*/
+
+#pragma once
+
+#include <vector>
+#include <unordered_map>
+#include <unordered_set>
+#include <stdint.h>
+#include <string>
+#include <iostream>
+#include <sstream>
+#include <algorithm>
+
+namespace claragenomics
+{
+
+using node_id_t     = int32_t;
+using edge_weight_t = int32_t;
+using edge_t        = std::pair<node_id_t, node_id_t>;
+
+/// \struct PairHash
+/// Hash function for a pair
+struct PairHash
+{
+public:
+    /// \brief Operator overload to define hash function
+    template <class T1, class T2>
+    size_t operator()(const std::pair<T1, T2>& pair) const
+    {
+        size_t hash_1 = std::hash<T1>()(pair.first);
+        size_t hash_2 = std::hash<T2>()(pair.second);
+        return hash_1 ^ hash_2;
+    }
+};
+
+/// \brief Object representing a generic graph structure
+class Graph
+{
+public:
+    /// \brief Default dtor
+    ~Graph() = default;
+
+    /// \brief Add edges to a graph
+    ///
+    /// \param node_id_from Source node ID
+    /// \param node_id_to Sink node ID
+    /// \param weight Edge weight
+    virtual void add_edge(node_id_t node_id_from, node_id_t node_id_to, edge_weight_t weight) = 0;
+
+    /// \brief Get a list of adjacent nodes to a given node
+    ///
+    /// \param node Node for which adjacent nodes are requested
+    /// \return Vector of adjacent node IDs
+    virtual const std::vector<node_id_t>& get_adjacent_nodes(node_id_t node) const
+    {
+        auto iter = adjacent_nodes_.find(node);
+        if (iter != adjacent_nodes_.end())
+        {
+            return iter->second;
+        }
+        else
+        {
+            return empty_;
+        }
+    }
+
+    /// \brief List all node IDs in the graph
+    ///
+    /// \return A vector of node IDs
+    virtual const std::vector<node_id_t> get_node_ids() const
+    {
+        std::vector<node_id_t> nodes;
+        for (auto iter : adjacent_nodes_)
+        {
+            nodes.push_back(iter.first);
+        }
+
+        return nodes;
+    }
+
+    /// \brief Get a list of all edges in the graph
+    ///
+    /// \return A vector of edges
+    virtual const std::vector<std::pair<edge_t, edge_weight_t>> get_edges() const
+    {
+        std::vector<std::pair<edge_t, edge_weight_t>> edges;
```

Comment:
Neat! Semantically better to have `begin(edges_)` or `edges_.begin()`?

Reply:
They are equivalent. `begin(x)` was introduced in C++11 and will call x.begin() in almost all cases.
I made it a habit to use it because it allows you to extend classes which don't have `.begin()` member functions without touching the class itself. You just need to add a `begin(const someclass& x)` function.